sql-plugin/src/main/330db/scala/org/apache/spark/sql/rapids/execution/GpuExecutorBroadcastHelper.scala

```diff
@@ -54,23 +70,50 @@ object GpuExecutorBroadcastHelper extends Arm {
    * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
    * fully coalesced ColumnarBatch loaded on to the GPU.
    *
-   * @param buildRelation - the executor broadcast produced by a shuffle exchange
+   * This method uses getSingleBatchWithVerification to handle the empty relation case 
+   * and the unexpected scenario of receiving more than a single batch
+   *
+   * @param shuffleData - the shuffle data from the executor for the "executor broadcast"
    * @param buildSchema - the schema expected for the output
    * @param buildOutput - the output attributes expected in case we receieve an empty RDD
    * @param metricsMap - metrics to populate while the shuffle coalesce iterator loads 
    *                     data from I/O to the device
    * @param targetSize - the target batch size in bytes for loading data on to the GPU 
    */
   def getExecutorBroadcastBatch(
-      buildRelation: RDD[ColumnarBatch], 
+      shuffleData: RDD[ColumnarBatch], 
       buildSchema: StructType,
       buildOutput: Seq[Attribute],
       metricsMap: Map[String, GpuMetric],
       targetSize: Long): ColumnarBatch = {
 
-    val it = shuffleCoalesceIterator(buildRelation, buildSchema, metricsMap, targetSize)
+    // TODO: Ensure we're not reading the shuffle data multiple times. 
+    // See https://github.com/NVIDIA/spark-rapids/issues/7599
+    val it = shuffleCoalesceIterator(shuffleData, buildSchema, metricsMap, targetSize)
     // In a broadcast hash join, right now the GPU expects a single batch
     ConcatAndConsumeAll.getSingleBatchWithVerification(it, buildOutput)
   }
 
+  /**
+   * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
+   * number of rows that the received batch contains
+   *
+   * The method uses shuffleDataIterator here to get the number of rows here without
+   * loading on to the GPU. The batch here needs to be closed.
+   *
+   */
+  def getExecutorBroadcastBatchNumRows(shuffleData: RDD[ColumnarBatch]): Int = {
+    // Ideally we cache this data so we're not reading the shuffle multiple times. 
+    // This requires caching the data and making it spillable/etc.
+    // See https://github.com/NVIDIA/spark-rapids/issues/7599
+    val it = shuffleDataIterator(shuffleData)
+    if (it.hasNext) {
+      withResource(it.next) { batch =>
+        batch.numRows
+      }
```

Comment:
Is it guaranteed the iterator produces all of the data in one iteration, or does this need to be a while loop?

Reply:
I'm not entirely sure, so I will convert it to a while loop.